src/net/ossl.cc

```diff
@@ -215,9 +217,59 @@ using pkcs12 = ssl_handle<PKCS12, PKCS12_free>;
 using ssl_ctx_ptr = ssl_handle<SSL_CTX, SSL_CTX_free>;
 using ssl_ptr = ssl_handle<SSL, SSL_free>;
 using x509_verify_param_ptr = ssl_handle<X509_VERIFY_PARAM, X509_VERIFY_PARAM_free>;
+using ssl_session_ptr = ssl_handle<SSL_SESSION, SSL_SESSION_free>;
+
+/**
+ * The purpose of this structure is to hold the AES and HMAC keys used to encrypt/decrypt TLS 1.3
+ * session tickets given to and presented by, the TLS client.
+ */
+struct session_ticket_keys {
+    std::array<uint8_t, 32> aes_key;
+    std::array<uint8_t, 32> hmac_key;
+
+    void generate_keys() {
+        generate_key(aes_key);
+        generate_key(hmac_key);
+    }
+
+    ~session_ticket_keys() {
+        clear_key(aes_key);
+        clear_key(hmac_key);
+    }
+
+private:
+    /**
+     * This will zeroize the key contents by writing 0xff, then 0x00,
+     * and finally 0xff over the memory that held the key
+     * @tparam N Size of the array
+     * @param key The key to clear
+     */
+    template<std::size_t N>
+    static void clear_key(std::array<uint8_t, N>& key) {
+        // Zeroize sensitive key data
+        memset(key.data(), 0xff, N);
+        memset(key.data(), 0x00, N);
+        memset(key.data(), 0xff, N);
+    }
```

Comment:
How does this custom implementation compare to `OPENSSL_cleanse`? Is there a reason to use one over the other?

Reply:
Looking at the source for `OPENSSL_cleanse`, it appears to ensure that the compiler doesn't attempt to optimize away the memset call, which is nice.  Whenever I've done key zeroization, I was told the 'proper' way was 0xff, 0x00, 0xff, but this may actually apply more to keys saved on disc rather than in memory.

Good catch on this, I"ll probably just switch over to that.